### PR TITLE
SF-SAPT

### PR DESCRIPTION
**Pull request by dgasmith**
## Description
Adds spin-flip SAPT to Psi4.

## Todos
Notable points that this PR has either accomplished or will accomplish.
* **User-Facing for Release Notes**
  - [x] Adds Spin-Flip SAPT for computing all multiplet interactions between two high-spin monomers.

## Checklist
- [x] Tests added for any new features
- [x] [All or relevant fraction of full tests run](http://psicode.org/psi4manual/master/build_planning.html#how-to-run-a-subset-of-tests)

## Status
- [x] Ready for review
- [x] Ready for merge

### CODE REVIEW

File: psi4/driver/procrouting/proc_table.py [hunk 1]
```diff
@@ -69,6 +69,7 @@
             'dfocc'         : proc.run_dfocc,  # full control over dfocc
             'qchf'          : proc.run_qchf,
             'ccd'           : proc.run_dfocc,
+            'sf-sapt'       : sapt.run_sf_sapt,
```
Comment (loriab) [inline psi4/driver/procrouting/proc_table.py]: `sf-sapt0`?

File: psi4/driver/procrouting/sapt/sapt_sf_terms.py [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+        if (Cr.shape[1] == 0) or (Cl.shape[1] == 0):
+            zero_append.append(num)
+            continue
+
+        if (rotation is not None) and (rotation[num] is not None):
+            mol = Cl.shape[1]
+            mor = Cr.shape[1]
+
+            if (rotation[num].shape[0] != mol) or (rotation[num].shape[1] != mor):
+                raise Exception("_sf_compute_JK: Tensor size does not match Cl (%d) /Cr (%d) : %s" %
```
Comment (loriab) [inline psi4/driver/procrouting/sapt/sapt_sf_terms.py]: ValidationError?

File: psi4/driver/procrouting/sapt/sapt_sf_terms.py [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+    Cright = [x[3] for x in pairs]
+
+    tmp_J, tmp_K = _sf_compute_JK(jk, Cleft, Cright, rotations)
+
+    J = {key: val for key, val in zip(names, tmp_J)}
+    K = {key: val for key, val in zip(names, tmp_K)}
+
+    ### Compute Terms
+    if do_print:
+        core.print_out("\n  ==> Computing Spin-Flip Exchange and Electostatics <== \n\n")
```
Comment (loriab) [inline psi4/driver/procrouting/sapt/sapt_sf_terms.py]: spelling in Elec

File: tests/sapt-sf1/input.dat [hunk 4]
```diff
@@ -0,0 +1,36 @@
+
```
Comment (loriab) [inline tests/sapt-sf1/input.dat]: Add a one-line descr. of test case for the docs starting with `#!`, please.

File: tests/sapt-sf1/input.dat [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+
+compare_dict = { #TEST
+    "SAPT ELST ENERGY":                 -0.02368051, #TEST
+    "E Elst10":                         -0.02368051, #TEST
+    "SAPT EXCH ENERGY":                  0.14186858, #TEST
+    "E Exch10(S^2) [highspin]":          0.14186858, #TEST
+    "E Exch10(S^2) [diagonal]":          0.00010209, #TEST
+    "E Exch10(S^2) [off-diagonal]":      0.14176648, #TEST
```
Comment (loriab) [inline tests/sapt-sf1/input.dat]: what would you think of `SAPT EXCH10(S^2),DIAGONAL ENERGY` or similar a la https://github.com/psi4/psi4/blob/master/psi4/driver/qcdb/psivardefs.py#L93?

Comment (dgasmith) [reply tests/sapt-sf1/input.dat]: Happy to do anything here, just trying out various modes.

File: psi4/driver/procrouting/proc_table.py (hunk 1 above)
Comment (dgasmith) [reply psi4/driver/procrouting/proc_table.py]: Well, its not really `sapt0` as you pointed out above. It is currently a bit too hard to integrate everything together without moving SAPT logic more py-side. Py-side SAPT is something that needs to be done for sure, but may be outside of scope for this PR.

**Comment by dgasmith** [comment]
Patched up, thanks for the comments.

File: psi4/driver/procrouting/proc_table.py (hunk 1 above)
Comment (andysim) [reply psi4/driver/procrouting/proc_table.py]: It's amazing how much functionality you were able to add in so few lines, even if the Py side SAPT is not in place.

**Pull request merged**